src/index.ts

```diff
@@ -1,87 +1,90 @@
-import { useState, useCallback, useEffect } from "react";
-import update from "immutability-helper";
+import {
+  useMemo,
+  useEffect,
+  useReducer,
+  useCallback,
+} from "react";
 
-import { HookArguments, HookReturnValues, Item } from "./types";
+import reducer from "./reducer";
+import {
+  HookArguments,
+  DefaultItem,
+  ActionType,
+  State,
+} from "./types";
+import { INITIAL_STATE } from "./constants";
 
-export const DEFAULT_ITEM_IDENTIFIER = "id";
-
-function useSelectedItems<T extends Record<any, any>>({
-  items = [],
-  itemIdentifier = DEFAULT_ITEM_IDENTIFIER,
-}: HookArguments<T>): HookReturnValues<T> {
-  const [selectedItems, setSelectedItems] = useState<T[]>([]);
-  const [itemsList, setItemsList] = useState<Item<T>[]>([]);
+function useSelectedItems<T extends DefaultItem, K>({
+  initialItems = [],
+  itemIdentifierKey,
+  initialSelectedItems = [],
+}: HookArguments<T, K>) {
+  const [items, dispatch] = useReducer(
+    reducer,
+    INITIAL_STATE,
+    (state: State) => ({
+      ...state,
+      itemIdentifierKey,
+    }),
+  );
 
   useEffect(() => {
-    const itemIdentifierIsValid = items.some((findItem: T) => (
-      findItem[itemIdentifier]
-    ));
+    const itemsWithInvalidIdentifers = initialItems.filter(
```

Comment:
I think you can use `some` like before, you just need to invert to `hasItemIdentifierInvalid` and change inside to use `!`

```ts
var hasItemIdentifierInvalid = items => items.some(item => !item.id)

hasItemIdentifierInvalid([{ id: 1 }]) // false
hasItemIdentifierInvalid([]) // false
hasItemIdentifierInvalid([{ name: 'gtkatakura' }]) // true
hasItemIdentifierInvalid([{ id: 1 }, { name: 'gtkatakura' }]) // true
hasItemIdentifierInvalid([{ id: 1 }, { id: 2, name: 'gtkatakura' }]) // false
```

And you don't need to check items.length because `some` already returns `false` by default when `items` are an empty array.

Reply:
Makes sense. To explain what I changed it to use the `filter` method: My goal was trying to check **all** the items. But by inverting the condition, as you showed in the snippet above, it also covers the case.

Reply:
Improved on [`ec4fb9d` (#11)](https://github.com/LauraBeatris/use-selected-items-hook/pull/11/commits/ec4fb9d4bb3be159be7e4bb54c2ac8936166785a)

Reply:
Improved on ec4fb9d (#11)